README.md
# \<tsante-mqtt\>

![Polymer](https://img.shields.io/badge/polymer-1.x-blue.svg)
[![Published on webcomponents.org](https://img.shields.io/badge/webcomponents.org-published-blue.svg)](https://www.webcomponents.org/element/telecomsante/tsante-mqtt)

A MQTT webcomponent, allowing to have several subscribers and publishers.

The component connect to the MQTT server through a websocket.

The component is based on the [Paho javascript client library](https://eclipse.org/paho/clients/js/).

The component has no GUI.

The component is provided in ES6

    <link rel="import" href="bower_components/tsante-mqtt.html">

or for older browser in ES5

    <link rel="import" href="bower_components/tsante-mqtt.es5.html">

## Quick example

> Nota : the demo doesn't work on webcomponents.org, due to permission access to
the localstorage, to see a demo please consult the demo link below.

```html
<tsante-mqtt host="ws://test.mosquitto.org:8080/" >
  <tsante-mqtt-subscriber topic="terminal/hello"></tsante-mqtt-subscriber>
  <tsante-mqtt-publisher id="publisher"
    topic="terminal/hello"
    payload="polymer"></tsante-mqtt-publisher>
</tsante-mqtt>
```

The component is licensed under the [ISC License](LICENSE.md)

Demo and doc are available on https://telecomsante.github.io/tsante-mqtt/

### version 1.4.0 and above

starting 1.4.0, it's possible to embed more deeper susbscribers and publishers

```html
<tsante-mqtt host="ws://test.mosquitto.org:8080/" >
  <div>
    <tsante-mqtt-subscriber topic="terminal/hello"></tsante-mqtt-subscriber>
    <tsante-mqtt-publisher id="publisher"
      topic="terminal/hello"
      payload="polymer"></tsante-mqtt-publisher>
  </div>
</tsante-mqtt>
```

> BREAKING CHANGE :
>
> `tsante-mqtt-received` events are fired by `tsante-mqtt-subscriber` elements rather than by `tsante-mqtt`
>
> however as the events bubble it's always possible to get the events from the `tsante-mqtt` ancestor. In this case some events can appear to be emitted several time, but look at the evt.target it must be different.


## Running tests

> __prerequisites :__ You will need to have __docker__ and __docker-compose__ installed

To run the test suites, you must start a local MQTT server. This is automatically done when running :

```bash
npm test
```

## npm scripts

__npm run mosquitto:start__ : starts a MQTT server into a docker container

__npm run mosquitto:stop__ : stops the MQTT server

__npm test__ : starts the MQTT server and runs the test suites on all installed browsers.

__npm run build-es5__ : builds es5 compatible components

## Git flow

The development of this component follows the git flow recommandations, so when you fork this repository and want to do a PR (push request), do it on the "develop" branch.


CHANGELOG.md
## 1.4.2

- Fix issue #15 : on changing the topic of a subscriber, it must unsubscribe the old topic, then subscribe to the now one.
- Fix issue #17 : Subscribers dynamically added to the the DOM must automatically subscribe

## 1.4.1

Fix issue #13

## 1.4.0

- allow tsante-subscribers to be embed deeper, they doesn't need to be direct child of tsante-mqtt

BREAKING CHANGE :

`tsante-mqtt-received` events are fired by `tsante-mqtt-subscriber` elements.

however as the events bubble it's always possible to get the events from the `tsante-mqtt` ancestor.
In this case some events can appear to be emitted several time, but look at the evt.target it must be different.

Internal :

- send tsante-mqtt's connected and client properties to all tsante-mqtt-subscriber children


## v1.3.0

- fix #9 : Add timeout and qos options for subscriber
  
## v1.2.0

- fix #7 : cannot publish the same payload when changing the topic

## v1.1.1

- fix #5 : enhance connection
- fix #4

## v1.1.0

- tsante-mqtt : send a `tsante-mqtt-connect-error` event on connection fail
- tsante-mqtt : retry connection when `retry` is set
- fix #2 : authenticated connection doesn't work
- fix #1 : add a ES5 version

## v1.0.1

- tsante-mqtt : reinitialize client when modifying host or clientID
- tsante-mqtt : on disconnect set connected to false
- tsante-mqtt-subscriber : when tsante-mqtt disconnect, unsubscribe the topic.

## v1.0.0

First public release


tsante-mqtt.js
Polymer({

  is: 'tsante-mqtt',

  properties: {
    /**
     * the url of the server
     *
     * by example :
     * `ws://test.mosquitto.org:8080/`
     *
     * @type {String}
     */
    host: {
      type: String,
    },

    /**
     * clientID
     *
     * the clientID is used by the server to retrieve the previous session
     *
     * @type {String}
     */
    clientID: {
      type: String,
      value: function () {
        return Math.random().toString(36).replace(/[^a-z]+/g, '')
      },
    },

    /**
     * username
     * @type {String}
     */
    username: {
      type: String,
    },

    /**
     * password
     * @type {String}
     */
    password: {
      type: String,
    },

    /**
     * timeout of the connection in seconds
     * @type {Number}
     */
    timeout: {
      type: Number,
      value: 10,
    },

    /**
     * retry connect after a period
     * the time is in seconds
     * @type {Number}
     */
    retry: {
      type: Number,
      value: 0,
    },

    /**
     * if true(default) the client and server persistent state
     * is deleted on successful connect.
     *
     * @type {Boolean}
     */
    cleanSession: {
      type: Boolean,
      value: true,
    },

    /**
     * message sent by the server when the client disconnects abnormally.
     *
     *
     * @type {String}
     */
    willMessage: {
      type: Object,
      value: function () { return {} },
    },

    /**
     * The mqtt client
     * @type {Object}
     */
    client: {
      type: Object,
    },

    /**
     * the connection status
     * @type {Boolean}
     */
    connected: {
      type: Boolean,
      value: false,
      readOnly: true,
      reflectToAttribute: true,
      observer: '_connectedChanged',
    },
  },

  observers: [
    '_sendInfoToPubSub(connected, client)',
    '_initializeClient(host,clientID)',
    'connect(username, password)',
  ],

  /**
   * on connected change or client change
   * update publishers and subscribers children
   * @method sendInfoToPubSub
   *
   * @param {*} connected
   * @param {*} client
   */
  _sendInfoToPubSub (connected, client) {
    let pubSubs = [].slice.call(this.querySelectorAll('tsante-mqtt-publisher, tsante-mqtt-subscriber'))
    for (let pubSub of pubSubs) {
      if (typeof pubSub.setNeededProperties === 'function') {
        pubSub.setNeededProperties(connected, client)
      }
    }
  },

  attached () {
    this.connect()
  },

  detached () {
    this.disconnect()
  },

  _initializeClient: function (host, clientID) {
    if (this.connected && this.client) { this.disconnect() }
    this.client = new Paho.MQTT.Client(this.host, this.clientID)
    this.client.onConnectionLost = this._onConnectionLost.bind(this)
    this.client.onMessageArrived = this._onMessageArrived.bind(this)
    this.client.onMessageDelivered = this._onMessageDelivered.bind(this)
    this.connect()
  },

  _onConnect: function () {
    this._setConnected(true)
  },

  /**
   * sent when the connection fails
   * @event tsante-mqtt-connect-error
   * @param  {String} message message of the error
   * @param  {Boolean} status status of the connection
   */
  _onFail: function () {
    this._setConnected(false)
    this.fire('tsante-mqtt-connect-error', {
      status: this.connected,
      message: 'fail to connect',
    })
    this._retry()
  },

  _retry: function () {
    if (this.retry) { setTimeout(this.connect.bind(this), this.retry * 1000) }
  },

  /**
   * Fired when the status of the connection change
   *
   * example of `evt.detail` :
   *
   * ```
   * { status: true }
   * ```
   *
   * @event tsante-mqtt-connect
   * @param {Boolean} status the status of the connection
   */
  _connectedChanged: function () {
    let subscribers = [].slice.call(this.querySelectorAll('tsante-mqtt-subscriber'))
    if (this.connected) {
      this.fire('tsante-mqtt-connect', { status: this.connected })
    }
    for (let subscriber of subscribers) {
      if (this.connected) {
        subscriber.subscribe(this.connected, this.client)
        subscriber._setSubscribed(true)
      // }else{
        // value._setSubscribed(false);
        // value.fire('tsante-mqtt-subscribed', { topic: this.topic, status: false });
      }
    }
  },

  _onConnectionLost: function (msg) {
    this._setConnected(false)
    this._retry()
  },

  /**
   * when a message is received transfert it to the subscribers
   *
   */
  _onMessageArrived: function (msg) {
    let subscribers = [].slice.call(this.querySelectorAll('tsante-mqtt-subscriber'))
    for (let subscriber of subscribers) {
      subscriber._received(msg)
    }
  },

  /**
   * called when a message has been delivered. All processing that this
   * Client will ever do has been completed.
   *
   * @method _onMessageDelivered
   * @param  {Object}            msg the delivered message
   */
  _onMessageDelivered: function (msg) {
    /**
     * fired when a message has been delivered
     *
     * example of `evt.detail` :
     * ```
     * {
     *    topic: 'terminal/hello',
     *    payload: 'world'
     * }
     * ```
     *
     * @event tsante-mqtt-delivered
     * @param {String} topic the topic
     * @param {String} payload the content of the message
     */
    if (typeof msg.payloadString === 'string') {
      this.fire('tsante-mqtt-delivered', { topic: msg.destinationName, payload: msg.payloadString })
    }
  },

  /**
   * open the connection to the server
   *
   * @method connect
   * @param  {String} username the login to use
   * @param  {String} password the password to use
   */
  connect: function (username, password) {
    if (!this.client) return
    this.debounce('connect', () => {
      const connectOption = {
        onSuccess: this._onConnect.bind(this),
        onFailure: this._onFail.bind(this),
        timeout: this.timeout,
        cleanSession: this.cleanSession,
      }
      if (this.username || username) {
        this.username = this.username || username
        connectOption.userName = this.username || username
      }
      if (this.password || password) {
        this.password = this.password || password
        connectOption.password = this.password || password
      }
      try {
        this.client.connect(connectOption)
      } catch (err) {
        this._onError(err.message)
      }
    }, 100)
  },

  /**
   * disconnect the client
   * @method disconnect
   */
  disconnect: function () {
    // if (!this.client) return;
    try {
      this.client.disconnect()
      this._setConnected(false)
      this.fire('tsante-mqtt-connect', { status: this.connected })
    } catch (err) {
      this._onError(err.message)
    }
  },

  /**
   * fired when an error occurs while connect or disconnect
   *
   * example of error :
   *  - "AMQJS0011E Invalid state not connecting or connected."
   *  - "AMQJS0011E Invalid state already connected."
   *
   * @event tsante-mqtt-error
   * @param  {String} errMessage message of the error
   */
  _onError: function (errMessage) {
    this.fire('tsante-mqtt-error', errMessage)
  },

})


tsante-mqtt-publisher.js
Polymer({
  is: 'tsante-mqtt-publisher',

  properties: {
    /**
     * the topic to publish in
     *
     * when the topic is changed, the payload is changed to null.
     *
     * @type {String}
     */
    topic: {
      type: String,
      observer: '_topicChanged',
    },
    /**
     * qos
     *
     * the value of the qos could be 0, 1, 2
     * @type {Number}
     */
    qos: {
      type: Number,
      value: 0,
    },
    /**
     * retain flag
     * @type {Boolean}
     */
    retained: {
      type: Boolean,
      value: false,
    },
    /**
     * the payload to send
     * @type {String}
     */
    payload: {
      type: String,
      value: null,
    },
    /**
     * the connection status of the `tsante-mqtt` ancestor
     * @type {Boolean}
     */
    _connected: {
      type: Boolean,
      value: false,
    },
    /**
     * the mqtt client of the `tsante-mqtt` ancestor
     * @type {Object}
     */
    _client: Object,
  },

  observers: [
    '_isConnected(_connected)',
    'publish(payload)',
  ],

  setNeededProperties: function (connected, client) {
    this._connected = connected
    this._client = client
  },

  _isConnected: function (_connected) {
    if (_connected && this._client) this.publish()
  },

  _topicChanged: function (newValue) {
    this.payload = null
    this.alreadyPublished = false
  },

  /**
   * publish to the server
   *
   * @method publish
   * @param  {String} payload the value to publish
   * @param  {Number} qos the qos of the message (0,1,2) by default this.qos
   * @param  {Boolean} retained flag indicates that the server must or not keep the value by default this.retained
   */
  publish: function (payload, qos, retained) {
    if (this._connected) {
      payload = typeof payload === 'string' ? payload : this.payload
      if (payload === null) return
      qos = (!qos || isNaN(qos)) ? this.qos : qos
      retained = typeof retained === 'boolean' ? retained : this.retained
      this._client.send(this.topic, payload, qos, retained)
    }
  },
})


tsante-mqtt-subscriber.js
Polymer({
  is: 'tsante-mqtt-subscriber',

  properties: {
    /**
     * topic to publish to
     * @type {String}
     */
    topic: {
      type: String,
      value: '#',
      observer: '_topicChanged',
    },
    /**
     * qos
     *
     * the value of the qos could be 0, 1, 2
     * https://www.hivemq.com/blog/mqtt-essentials-part-6-mqtt-quality-of-service-levels
     * @type {Number}
     */
    qos: {
      type: Number,
      value: null,
    },
    /**
     * timeout
     *
     * the timeout value in seconds
     * @type {Number}
     */
    timeout: {
      type: Number,
      value: null,
    },

    /**
     * set when subscribe is sucessful
     * @type {Boolean}
     */
    subscribed: {
      type: Boolean,
      value: false,
      readOnly: true,
      reflectToAttribute: true,
    },

    _connected: {
      type: Boolean,
      observer: 'isConnected',
    },
    _client: Object,
  },

  attached: function () {
    let parent = this.parentElement
    while (['tsante-mqtt', 'body'].indexOf(parent.tagName.toLowerCase()) === -1) {
      parent = parent.parentElement
    }
    this.subscribe(parent.connected, parent.client)
  },

  detached: function () {
    this.unsubscribe(this.topic)
  },

  isConnected: function () {
    if (!this._connected && this.subscribed) this.unsubscribe()
  },

  /**
   * filter received messages
   *
   * This method is called by the `tsante-mqtt` ancestor when a message is received
   *
   *
   * @param {destinationName, payloadString} msg
   */
  _received: function (msg) {
    let rightTopic
    if (this.topic[this.topic.length - 1] === '#') {
      let shortTopic = this.topic.split('/').slice(0, -1).join('')
      rightTopic = msg.destinationName.split('/').slice(0, -1).join('').slice(0, shortTopic.length) === shortTopic
    }
    if (rightTopic || msg.destinationName === this.topic) {
      this._fireReceived({ topic: msg.destinationName, payload: msg.payloadString })
    }
  },

  /**
   * fire an event on received message
   *
   * example of the `evt.detail` :
   *
   * ```
   * {
   *   topic: "terminal/hello",
   *   payload: "polymer"
   * }
   * ```
   *
   * @event tsante-mqtt-received
   * @param  {String} topic topic of the received message
   * @param  {String} payload content of the received message
   */
  _fireReceived: function ({topic, payload}) {
    this.fire('tsante-mqtt-received', { topic, payload })
  },

  setNeededProperties: function (connected) {
    this._connected = connected
  },

  /**
   * When topic change, if the component is subscribed
   * then unsubscribe the old topic before subscribing to the new one
   *
   * @method _topicChanged
   * @param  {String}  newValue the new topic to subscribe to
   * @param  {String}  oldValue the previous topic
   */
  _topicChanged: function (newValue, oldValue) {
    if (newValue !== oldValue) {
      if (oldValue && this.subscribed) {
        this.addEventListener('tsante-mqtt-subscribed', this.subscribe, {once: true})
        this.unsubscribe(oldValue)
      } else {
        this.subscribe()
      }
    }
  },

  /**
   * subscribe to the topic
   * this method is called by the `tsante-mqtt`
   * @method subscribe
   *
   * @param {*} connected the connected status of the tsante-mqtt parent
   * @param {*} client the mqtt client of the tsante-mqtt parent
   */
  subscribe: function (connected, client) {
    this._client = this._client || client
    this._connected = this._connected || connected
    if (this._connected && !this.subscribed) {
      this._client.subscribe(this.topic, {
        onSuccess: this._onSubscribe.bind(this),
        onFailure: this._onSubscribeFail.bind(this),
        invocationContext: { topic: this.topic },
      })
    };
  },

  /**
   * fired when subscribed to a topic
   *
   * example of `evt.detail` :
   *
   * ```
   * {
   *    topic: "terminal/hello",
   *    status: true
   * }
   * ```
   *
   * @event tsante-mqtt-subscribed
   * @param  {String}  topic the topic on which we subscribe
   * @param  {Boolean}  status the subscription status
   */
  _onSubscribe: function (evt) {
    this._setSubscribed(true)
    this.fire('tsante-mqtt-subscribed', { topic: evt.invocationContext.topic, status: true })
  },

  _onSubscribeFail: function (evt) {
    this._setSubscribed(false)
    this._onError(evt)
  },

  /**
   * unsubscribe of the given topic
   * @method unsubscribe
   * @param  {String}  topic the topic to unsubscribe (default this.topic )
   */
  unsubscribe: function (topic = this.topic) {
    this._setSubscribed(false)
    this.fire('tsante-mqtt-subscribed', { topic: topic, status: false })
  },

  /**
   * fired on subscription/unsupscription error
   * @event tsante-mqtt-error
   * @param  {String}  message the error message
   */
  _onError (err) {
    this.fire('tsante-mqtt-error', err.message)
  },
})


tsante-mqtt-subscriber.es5.js
'use strict';

Polymer({
  is: 'tsante-mqtt-subscriber',

  properties: {
    /**
     * topic to publish to
     * @type {String}
     */
    topic: {
      type: String,
      value: '#',
      observer: '_topicChanged'
    },
    /**
     * qos
     *
     * the value of the qos could be 0, 1, 2
     * https://www.hivemq.com/blog/mqtt-essentials-part-6-mqtt-quality-of-service-levels
     * @type {Number}
     */
    qos: {
      type: Number,
      value: null
    },
    /**
     * timeout
     *
     * the timeout value in seconds
     * @type {Number}
     */
    timeout: {
      type: Number,
      value: null
    },

    /**
     * set when subscribe is sucessful
     * @type {Boolean}
     */
    subscribed: {
      type: Boolean,
      value: false,
      readOnly: true,
      reflectToAttribute: true
    },

    _connected: {
      type: Boolean,
      observer: 'isConnected'
    },
    _client: Object
  },

  attached: function attached() {
    var parent = this.parentElement;
    while (['tsante-mqtt', 'body'].indexOf(parent.tagName.toLowerCase()) === -1) {
      parent = parent.parentElement;
    }
    this.subscribe(parent.connected, parent.client);
  },

  detached: function detached() {
    this.unsubscribe(this.topic);
  },

  isConnected: function isConnected() {
    if (!this._connected && this.subscribed) this.unsubscribe();
  },

  /**
   * filter received messages
   *
   * This method is called by the `tsante-mqtt` ancestor when a message is received
   *
   *
   * @param {destinationName, payloadString} msg
   */
  _received: function _received(msg) {
    var rightTopic = void 0;
    if (this.topic[this.topic.length - 1] === '#') {
      var shortTopic = this.topic.split('/').slice(0, -1).join('');
      rightTopic = msg.destinationName.split('/').slice(0, -1).join('').slice(0, shortTopic.length) === shortTopic;
    }
    if (rightTopic || msg.destinationName === this.topic) {
      this._fireReceived({ topic: msg.destinationName, payload: msg.payloadString });
    }
  },

  /**
   * fire an event on received message
   *
   * example of the `evt.detail` :
   *
   * ```
   * {
   *   topic: "terminal/hello",
   *   payload: "polymer"
   * }
   * ```
   *
   * @event tsante-mqtt-received
   * @param  {String} topic topic of the received message
   * @param  {String} payload content of the received message
   */
  _fireReceived: function _fireReceived(_ref) {
    var topic = _ref.topic,
        payload = _ref.payload;

    this.fire('tsante-mqtt-received', { topic: topic, payload: payload });
  },

  setNeededProperties: function setNeededProperties(connected) {
    this._connected = connected;
  },

  /**
   * When topic change, if the component is subscribed
   * then unsubscribe the old topic before subscribing to the new one
   *
   * @method _topicChanged
   * @param  {String}  newValue the new topic to subscribe to
   * @param  {String}  oldValue the previous topic
   */
  _topicChanged: function _topicChanged(newValue, oldValue) {
    if (newValue !== oldValue) {
      if (oldValue && this.subscribed) {
        this.addEventListener('tsante-mqtt-subscribed', this.subscribe, { once: true });
        this.unsubscribe(oldValue);
      } else {
        this.subscribe();
      }
    }
  },

  /**
   * subscribe to the topic
   * this method is called by the `tsante-mqtt`
   * @method subscribe
   *
   * @param {*} connected the connected status of the tsante-mqtt parent
   * @param {*} client the mqtt client of the tsante-mqtt parent
   */
  subscribe: function subscribe(connected, client) {
    this._client = this._client || client;
    this._connected = this._connected || connected;
    if (this._connected && !this.subscribed) {
      this._client.subscribe(this.topic, {
        onSuccess: this._onSubscribe.bind(this),
        onFailure: this._onSubscribeFail.bind(this),
        invocationContext: { topic: this.topic }
      });
    };
  },

  /**
   * fired when subscribed to a topic
   *
   * example of `evt.detail` :
   *
   * ```
   * {
   *    topic: "terminal/hello",
   *    status: true
   * }
   * ```
   *
   * @event tsante-mqtt-subscribed
   * @param  {String}  topic the topic on which we subscribe
   * @param  {Boolean}  status the subscription status
   */
  _onSubscribe: function _onSubscribe(evt) {
    this._setSubscribed(true);
    this.fire('tsante-mqtt-subscribed', { topic: evt.invocationContext.topic, status: true });
  },

  _onSubscribeFail: function _onSubscribeFail(evt) {
    this._setSubscribed(false);
    this._onError(evt);
  },

  /**
   * unsubscribe of the given topic
   * @method unsubscribe
   * @param  {String}  topic the topic to unsubscribe (default this.topic )
   */
  unsubscribe: function unsubscribe() {
    var topic = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : this.topic;

    this._setSubscribed(false);
    this.fire('tsante-mqtt-subscribed', { topic: topic, status: false });
  },

  /**
   * fired on subscription/unsupscription error
   * @event tsante-mqtt-error
   * @param  {String}  message the error message
   */
  _onError: function _onError(err) {
    this.fire('tsante-mqtt-error', err.message);
  }
});


components/tsante-mqtt/tsante-mqtt-publisher.es5.js
'use strict';

Polymer({
  is: 'tsante-mqtt-publisher',

  properties: {
    /**
     * the topic to publish in
     *
     * when the topic is changed, the payload is changed to null.
     *
     * @type {String}
     */
    topic: {
      type: String,
      observer: '_topicChanged'
    },
    /**
     * qos
     *
     * the value of the qos could be 0, 1, 2
     * @type {Number}
     */
    qos: {
      type: Number,
      value: 0
    },
    /**
     * retain flag
     * @type {Boolean}
     */
    retained: {
      type: Boolean,
      value: false
    },
    /**
     * the payload to send
     * @type {String}
     */
    payload: {
      type: String,
      value: null
    },
    /**
     * the connection status of the `tsante-mqtt` ancestor
     * @type {Boolean}
     */
    _connected: {
      type: Boolean,
      value: false
    },
    /**
     * the mqtt client of the `tsante-mqtt` ancestor
     * @type {Object}
     */
    _client: Object
  },

  observers: ['_isConnected(_connected)', 'publish(payload)'],

  setNeededProperties: function setNeededProperties(connected, client) {
    this._connected = connected;
    this._client = client;
  },

  _isConnected: function _isConnected(_connected) {
    if (_connected && this._client) this.publish();
  },

  _topicChanged: function _topicChanged(newValue) {
    this.payload = null;
    this.alreadyPublished = false;
  },

  /**
   * publish to the server
   *
   * @method publish
   * @param  {String} payload the value to publish
   * @param  {Number} qos the qos of the message (0,1,2) by default this.qos
   * @param  {Boolean} retained flag indicates that the server must or not keep the value by default this.retained
   */
  publish: function publish(payload, qos, retained) {
    if (this._connected) {
      payload = typeof payload === 'string' ? payload : this.payload;
      if (payload === null) return;
      qos = !qos || isNaN(qos) ? this.qos : qos;
      retained = typeof retained === 'boolean' ? retained : this.retained;
      this._client.send(this.topic, payload, qos, retained);
    }
  }
});


tsante-mqtt.es5.js
'use strict';

Polymer({

  is: 'tsante-mqtt',

  properties: {
    /**
     * the url of the server
     *
     * by example :
     * `ws://test.mosquitto.org:8080/`
     *
     * @type {String}
     */
    host: {
      type: String
    },

    /**
     * clientID
     *
     * the clientID is used by the server to retrieve the previous session
     *
     * @type {String}
     */
    clientID: {
      type: String,
      value: function value() {
        return Math.random().toString(36).replace(/[^a-z]+/g, '');
      }
    },

    /**
     * username
     * @type {String}
     */
    username: {
      type: String
    },

    /**
     * password
     * @type {String}
     */
    password: {
      type: String
    },

    /**
     * timeout of the connection in seconds
     * @type {Number}
     */
    timeout: {
      type: Number,
      value: 10
    },

    /**
     * retry connect after a period
     * the time is in seconds
     * @type {Number}
     */
    retry: {
      type: Number,
      value: 0
    },

    /**
     * if true(default) the client and server persistent state
     * is deleted on successful connect.
     *
     * @type {Boolean}
     */
    cleanSession: {
      type: Boolean,
      value: true
    },

    /**
     * message sent by the server when the client disconnects abnormally.
     *
     *
     * @type {String}
     */
    willMessage: {
      type: Object,
      value: function value() {
        return {};
      }
    },

    /**
     * The mqtt client
     * @type {Object}
     */
    client: {
      type: Object
    },

    /**
     * the connection status
     * @type {Boolean}
     */
    connected: {
      type: Boolean,
      value: false,
      readOnly: true,
      reflectToAttribute: true,
      observer: '_connectedChanged'
    }
  },

  observers: ['_sendInfoToPubSub(connected, client)', '_initializeClient(host,clientID)', 'connect(username, password)'],

  /**
   * on connected change or client change
   * update publishers and subscribers children
   * @method sendInfoToPubSub
   *
   * @param {*} connected
   * @param {*} client
   */
  _sendInfoToPubSub: function _sendInfoToPubSub(connected, client) {
    var pubSubs = [].slice.call(this.querySelectorAll('tsante-mqtt-publisher, tsante-mqtt-subscriber'));
    var _iteratorNormalCompletion = true;
    var _didIteratorError = false;
    var _iteratorError = undefined;

    try {
      for (var _iterator = pubSubs[Symbol.iterator](), _step; !(_iteratorNormalCompletion = (_step = _iterator.next()).done); _iteratorNormalCompletion = true) {
        var pubSub = _step.value;

        if (typeof pubSub.setNeededProperties === 'function') {
          pubSub.setNeededProperties(connected, client);
        }
      }
    } catch (err) {
      _didIteratorError = true;
      _iteratorError = err;
    } finally {
      try {
        if (!_iteratorNormalCompletion && _iterator.return) {
          _iterator.return();
        }
      } finally {
        if (_didIteratorError) {
          throw _iteratorError;
        }
      }
    }
  },
  attached: function attached() {
    this.connect();
  },
  detached: function detached() {
    this.disconnect();
  },


  _initializeClient: function _initializeClient(host, clientID) {
    if (this.connected && this.client) {
      this.disconnect();
    }
    this.client = new Paho.MQTT.Client(this.host, this.clientID);
    this.client.onConnectionLost = this._onConnectionLost.bind(this);
    this.client.onMessageArrived = this._onMessageArrived.bind(this);
    this.client.onMessageDelivered = this._onMessageDelivered.bind(this);
    this.connect();
  },

  _onConnect: function _onConnect() {
    this._setConnected(true);
  },

  /**
   * sent when the connection fails
   * @event tsante-mqtt-connect-error
   * @param  {String} message message of the error
   * @param  {Boolean} status status of the connection
   */
  _onFail: function _onFail() {
    this._setConnected(false);
    this.fire('tsante-mqtt-connect-error', {
      status: this.connected,
      message: 'fail to connect'
    });
    this._retry();
  },

  _retry: function _retry() {
    if (this.retry) {
      setTimeout(this.connect.bind(this), this.retry * 1000);
    }
  },

  /**
   * Fired when the status of the connection change
   *
   * example of `evt.detail` :
   *
   * ```
   * { status: true }
   * ```
   *
   * @event tsante-mqtt-connect
   * @param {Boolean} status the status of the connection
   */
  _connectedChanged: function _connectedChanged() {
    var subscribers = [].slice.call(this.querySelectorAll('tsante-mqtt-subscriber'));
    if (this.connected) {
      this.fire('tsante-mqtt-connect', { status: this.connected });
    }
    var _iteratorNormalCompletion2 = true;
    var _didIteratorError2 = false;
    var _iteratorError2 = undefined;

    try {
      for (var _iterator2 = subscribers[Symbol.iterator](), _step2; !(_iteratorNormalCompletion2 = (_step2 = _iterator2.next()).done); _iteratorNormalCompletion2 = true) {
        var subscriber = _step2.value;

        if (this.connected) {
          subscriber.subscribe(this.connected, this.client);
          subscriber._setSubscribed(true);
          // }else{
          // value._setSubscribed(false);
          // value.fire('tsante-mqtt-subscribed', { topic: this.topic, status: false });
        }
      }
    } catch (err) {
      _didIteratorError2 = true;
      _iteratorError2 = err;
    } finally {
      try {
        if (!_iteratorNormalCompletion2 && _iterator2.return) {
          _iterator2.return();
        }
      } finally {
        if (_didIteratorError2) {
          throw _iteratorError2;
        }
      }
    }
  },

  _onConnectionLost: function _onConnectionLost(msg) {
    this._setConnected(false);
    this._retry();
  },

  /**
   * when a message is received transfert it to the subscribers
   *
   */
  _onMessageArrived: function _onMessageArrived(msg) {
    var subscribers = [].slice.call(this.querySelectorAll('tsante-mqtt-subscriber'));
    var _iteratorNormalCompletion3 = true;
    var _didIteratorError3 = false;
    var _iteratorError3 = undefined;

    try {
      for (var _iterator3 = subscribers[Symbol.iterator](), _step3; !(_iteratorNormalCompletion3 = (_step3 = _iterator3.next()).done); _iteratorNormalCompletion3 = true) {
        var subscriber = _step3.value;

        subscriber._received(msg);
      }
    } catch (err) {
      _didIteratorError3 = true;
      _iteratorError3 = err;
    } finally {
      try {
        if (!_iteratorNormalCompletion3 && _iterator3.return) {
          _iterator3.return();
        }
      } finally {
        if (_didIteratorError3) {
          throw _iteratorError3;
        }
      }
    }
  },

  /**
   * called when a message has been delivered. All processing that this
   * Client will ever do has been completed.
   *
   * @method _onMessageDelivered
   * @param  {Object}            msg the delivered message
   */
  _onMessageDelivered: function _onMessageDelivered(msg) {
    /**
     * fired when a message has been delivered
     *
     * example of `evt.detail` :
     * ```
     * {
     *    topic: 'terminal/hello',
     *    payload: 'world'
     * }
     * ```
     *
     * @event tsante-mqtt-delivered
     * @param {String} topic the topic
     * @param {String} payload the content of the message
     */
    if (typeof msg.payloadString === 'string') {
      this.fire('tsante-mqtt-delivered', { topic: msg.destinationName, payload: msg.payloadString });
    }
  },

  /**
   * open the connection to the server
   *
   * @method connect
   * @param  {String} username the login to use
   * @param  {String} password the password to use
   */
  connect: function connect(username, password) {
    var _this = this;

    if (!this.client) return;
    this.debounce('connect', function () {
      var connectOption = {
        onSuccess: _this._onConnect.bind(_this),
        onFailure: _this._onFail.bind(_this),
        timeout: _this.timeout,
        cleanSession: _this.cleanSession
      };
      if (_this.username || username) {
        _this.username = _this.username || username;
        connectOption.userName = _this.username || username;
      }
      if (_this.password || password) {
        _this.password = _this.password || password;
        connectOption.password = _this.password || password;
      }
      try {
        _this.client.connect(connectOption);
      } catch (err) {
        _this._onError(err.message);
      }
    }, 100);
  },

  /**
   * disconnect the client
   * @method disconnect
   */
  disconnect: function disconnect() {
    // if (!this.client) return;
    try {
      this.client.disconnect();
      this._setConnected(false);
      this.fire('tsante-mqtt-connect', { status: this.connected });
    } catch (err) {
      this._onError(err.message);
    }
  },

  /**
   * fired when an error occurs while connect or disconnect
   *
   * example of error :
   *  - "AMQJS0011E Invalid state not connecting or connected."
   *  - "AMQJS0011E Invalid state already connected."
   *
   * @event tsante-mqtt-error
   * @param  {String} errMessage message of the error
   */
  _onError: function _onError(errMessage) {
    this.fire('tsante-mqtt-error', errMessage);
  }

});
